src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+           ↓
+       01101000     x
+       01100111     x - 1
+       --------
+       01100000     x .&. (x - 1)
+     -}
+    go w = if w .&. (w - 1) == 0 then w else go (w .&. (w - 1))
+
+data MerkleProof a =
+  MerkleProof { getMerkleProof :: [ProofElem a]  -- ^ list of proof elements
+              , getMerkleProofRoot :: Maybe (SizedMerkleRoot a) -- ^ leaf root proof is
+                                                                -- constructed for
+              }
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+data ProofElem a = ProofElem
+  { nodeRoot    :: SizedMerkleRoot a
+  , siblingRoot :: SizedMerkleRoot a
+  , nodeSide    :: Side
+  } deriving (Show, Eq, Ord, Generic, Serialise)
+
+data Side = L | R
+  deriving (Show, Eq, Ord, Generic, Serialise)
+
+-- | Construct merkle proof by recusive walking the tree and collecting ProofElem until we hit
+-- matching leafRoot
+mkMerkleProof :: forall a. SizedMerkleTree a -- ^ merkle tree we want to
+                                             -- construct a proof from
+                        -> LeafIndex         -- ^ leaf index used for proof
+                        -> MerkleProof a
+mkMerkleProof SizedMerkleEmpty _ = MerkleProof [] Nothing
+mkMerkleProof (SizedMerkleTree rootNode) n =
+  let (path, proofLeaf) = constructPath [] rootNode
+  in MerkleProof path proofLeaf
+  where
+    constructPath :: [ProofElem a] -> SizedMerkleNode a -> ([ProofElem a], Maybe (SizedMerkleRoot a))
+    constructPath pElems (SizedMerkleLeaf leafRoot' leafIndex _)
+      | n == leafIndex = (pElems, Just leafRoot')
+      | otherwise = ([], Nothing)
+    constructPath pElems (SizedMerkleBranch bRoot ln rn) =
+         case (lPath, rPath) of
+              ((xs, Nothing), (ys, Nothing)) -> (xs ++ ys, Nothing)
+              ((xs, Just l), (ys, _))        -> (xs ++ ys, Just l)
+              ((xs, _), (ys, Just r))        -> (xs ++ ys, Just r)
+      where
+        lProofElem = ProofElem (mRoot ln) (mRoot rn) L
+        rProofElem = ProofElem (mRoot rn) (mRoot ln) R
+
+        lPath = constructPath (lProofElem:pElems) ln
+        rPath = constructPath (rProofElem:pElems) rn
+
+-- | Validate a merkle tree proof
+validateMerkleProof :: forall a. MerkleProof a -> SizedMerkleTree a -> Bool
+validateMerkleProof (MerkleProof proofElems (Just leafRoot)) (SizedMerkleTree treeNode) =
+    validate proofElems leafRoot
+  where
+    validate :: [ProofElem a] -> SizedMerkleRoot a -> Bool
+    validate [] proofRoot = proofRoot == mRoot treeNode
+    validate (pElem:pElems) proofRoot
+      | proofRoot /= nodeRoot pElem = False
+      | otherwise = validate pElems $ hashProofElem pElem
+
+    hashProofElem :: ProofElem a -> SizedMerkleRoot a
+    hashProofElem (ProofElem pRoot sibRoot side) =
+      case side of
+        L -> mkBranchRootHash pRoot sibRoot
+        R -> mkBranchRootHash sibRoot pRoot
+
+validateMerkleProof (MerkleProof _ Nothing) _ = False
+
+
+-- | TODO, create tests and put in test module
```

Comment:
I'd recommend putting the tests in separate `./test/` folder and using quickcheck/smallcheck/hedgehog and some test frameworks (you'd better ask @flyingleafe what test framework to use).

I also recommend having a REPL.hs file so you can rearrange situation from the test to play with it, since the test engine will generate random tree each time.

Reply:
Why do you need a `REPL.hs` when you have `stack ghci`, @Heimdell?

Reply:
Otherwise I agree, @lambda76 you better create a test module (similar to existing ones in `test` subdir) using Tasty + hspec from the very beginning.

Reply:
ok